File: scripts/fallback_client.py
"""
Hermes Model Fallback Client (Python)
--------------------------------------
Automatically retries and switches AI models on HTTP 429, 503, or Timeout errors.

Usage:
    from fallback_client import FallbackClient

    client = FallbackClient("config.json")
    response = client.chat("Your prompt here")
    print(response)
"""

import json
import os
import time
import requests
from typing import Optional


# ── Errors that trigger a retry / fallback ─────────────────────────────────────
RETRYABLE_STATUS = {429, 503}


class AllModelsFailedError(Exception):
    """Raised when every model in the list has been exhausted."""
    pass


# ── Provider adapters ──────────────────────────────────────────────────────────

def _resolve_key(raw_key: str) -> str:
    """Supports 'ENV:VAR_NAME' syntax to read from environment variables."""
    if raw_key.startswith("ENV:"):
        env_var = raw_key[4:]
        value = os.environ.get(env_var, "")
        if not value:
            raise EnvironmentError(f"Environment variable '{env_var}' is not set.")
        return value
    return raw_key


def _call_gemini(model: str, api_key: str, prompt: str, timeout: int) -> str:
    url = (
        f"https://generativelanguage.googleapis.com/v1beta/models/"
        f"{model}:generateContent?key={api_key}"
    )
    body = {"contents": [{"parts": [{"text": prompt}]}]}
    resp = requests.post(url, json=body, timeout=timeout)
    resp.raise_for_status()
    return resp.json()["candidates"][0]["content"]["parts"][0]["text"]


def _call_openai(model: str, api_key: str, prompt: str, timeout: int) -> str:
    url = "https://api.openai.com/v1/chat/completions"
    headers = {"Authorization": f"Bearer {api_key}", "Content-Type": "application/json"}
    body = {"model": model, "messages": [{"role": "user", "content": prompt}]}
    resp = requests.post(url, headers=headers, json=body, timeout=timeout)
    resp.raise_for_status()
    return resp.json()["choices"][0]["message"]["content"]


def _call_claude(model: str, api_key: str, prompt: str, timeout: int) -> str:
    url = "https://api.anthropic.com/v1/messages"
    headers = {
        "x-api-key": api_key,
        "anthropic-version": "2023-06-01",
        "Content-Type": "application/json",
    }
    body = {
        "model": model,
        "max_tokens": 1024,
        "messages": [{"role": "user", "content": prompt}],
    }
    resp = requests.post(url, headers=headers, json=body, timeout=timeout)
    resp.raise_for_status()
    return resp.json()["content"][0]["text"]


def _call_mistral(model: str, api_key: str, prompt: str, timeout: int) -> str:
    url = "https://api.mistral.ai/v1/chat/completions"
    headers = {"Authorization": f"Bearer {api_key}", "Content-Type": "application/json"}
    body = {"model": model, "messages": [{"role": "user", "content": prompt}]}
    resp = requests.post(url, headers=headers, json=body, timeout=timeout)
    resp.raise_for_status()
    return resp.json()["choices"][0]["message"]["content"]


def _call_groq(model: str, api_key: str, prompt: str, timeout: int) -> str:
    url = "https://api.groq.com/openai/v1/chat/completions"
    headers = {"Authorization": f"Bearer {api_key}", "Content-Type": "application/json"}
    body = {"model": model, "messages": [{"role": "user", "content": prompt}]}
    resp = requests.post(url, headers=headers, json=body, timeout=timeout)
    resp.raise_for_status()
    return resp.json()["choices"][0]["message"]["content"]


def _call_ollama(model: str, api_key: str, prompt: str, timeout: int) -> str:
    """Local Ollama — no API key needed."""
    url = "http://localhost:11434/api/generate"
    body = {"model": model, "prompt": prompt, "stream": False}
    resp = requests.post(url, json=body, timeout=timeout)
    resp.raise_for_status()
    return resp.json()["response"]


PROVIDERS = {
    "gemini": _call_gemini,
    "openai": _call_openai,
    "claude": _call_claude,
    "mistral": _call_mistral,
    "groq": _call_groq,
    "ollama": _call_ollama,
}


# ── Main Client ────────────────────────────────────────────────────────────────

class FallbackClient:
    """
    Resilient AI client that automatically falls back to the next model
    on HTTP 429, 503, or Timeout errors.

    Parameters
    ----------
    config_path : str
        Path to your config.json file.
    verbose : bool
        Print which model is being used / when a fallback happens.
    """

    def __init__(self, config_path: str = "config.json", verbose: bool = True):
        with open(config_path, "r") as f:
            cfg = json.load(f)

        self.models = cfg["models"]
        self.retry_limit = cfg.get("retry_limit", 2)
        self.timeout = cfg.get("timeout_seconds", 30)
        self.verbose = verbose

    def _log(self, msg: str):
        if self.verbose:
            print(f"[HermesFallback] {msg}")

    def _try_model(self, model_cfg: dict, prompt: str) -> Optional[str]:
        """
        Attempt a single model up to (retry_limit + 1) times.
        Returns the response string, or None if all attempts failed.
        """
        provider = model_cfg["provider"]
        model = model_cfg["model"]
        api_key = _resolve_key(model_cfg.get("api_key", ""))
        caller = PROVIDERS.get(provider)

        if caller is None:
            self._log(f"Unknown provider '{provider}', skipping.")
            return None

        for attempt in range(1, self.retry_limit + 2):  # e.g. 1, 2, 3
            try:
                self._log(f"Trying {provider}/{model} (attempt {attempt})...")
                result = caller(model, api_key, prompt, self.timeout)
                self._log(f"✅ Success with {provider}/{model}")
                return result

            except requests.exceptions.Timeout:
                self._log(f"⏱ Timeout on {provider}/{model} (attempt {attempt})")

            except requests.exceptions.HTTPError as e:
                status = e.response.status_code if e.response is not None else 0
                if status in RETRYABLE_STATUS:
                    self._log(f"⚠️  HTTP {status} on {provider}/{model} (attempt {attempt})")
                    # Brief back-off before retry
                    if attempt <= self.retry_limit:
                        time.sleep(2 ** (attempt - 1))  # 1s, 2s
                else:
                    # Non-retryable error (e.g. 401, 400) — skip this model immediately
                    self._log(f"❌ Non-retryable HTTP {status} on {provider}/{model}: {e}")
                    return None

            except requests.exceptions.ConnectionError as e:
                self._log(f"🔌 Connection error on {provider}/{model}: {e}")

        self._log(f"❌ All attempts exhausted for {provider}/{model}, switching...")
        return None

    def chat(self, prompt: str) -> str:
        """
        Send a prompt. Automatically falls back through all configured models.

        Returns
        -------
        str
            The first successful response text.

        Raises
        ------
        AllModelsFailedError
            If every model fails after all retries.
        """
        for model_cfg in self.models:
            result = self._try_model(model_cfg, prompt)
            if result is not None:
                return result

        raise AllModelsFailedError(
            "All models failed (429 / 503 / Timeout). "
            "Check your API keys and quotas, or add more fallback models in config.json."
        )


# ── Quick test ─────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    client = FallbackClient("config.json")
    answer = client.chat("Say hello in one sentence.")
    print("Response:", answer)

File: scripts/fallbackClient.js
/**
 * Hermes Model Fallback Client (JavaScript / Node.js)
 * -----------------------------------------------------
 * Automatically retries and switches AI models on HTTP 429, 503, or Timeout errors.
 *
 * Usage:
 *   const { FallbackClient } = require('./fallbackClient');
 *   const client = new FallbackClient('./config.json');
 *   const response = await client.chat('Your prompt here');
 *   console.log(response);
 *
 * Requirements:
 *   npm install node-fetch   (Node 16 and below)
 *   Node 18+ has built-in fetch — no install needed.
 */

const fs = require('fs');
const path = require('path');

// Use built-in fetch (Node 18+) or node-fetch
let fetchFn;
try {
  fetchFn = fetch; // Node 18+ global
} catch {
  fetchFn = require('node-fetch');
}

// ── Errors that trigger a retry / fallback ─────────────────────────────────────
const RETRYABLE_STATUS = new Set([429, 503]);

class AllModelsFailedError extends Error {
  constructor(message) {
    super(message);
    this.name = 'AllModelsFailedError';
  }
}

// ── Helpers ────────────────────────────────────────────────────────────────────

/** Supports 'ENV:VAR_NAME' syntax to read from environment variables. */
function resolveKey(rawKey) {
  if (!rawKey) return '';
  if (rawKey.startsWith('ENV:')) {
    const envVar = rawKey.slice(4);
    const value = process.env[envVar] || '';
    if (!value) throw new Error(`Environment variable '${envVar}' is not set.`);
    return value;
  }
  return rawKey;
}

/** Simple timeout wrapper around fetch */
async function fetchWithTimeout(url, options, timeoutMs) {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const res = await fetchFn(url, { ...options, signal: controller.signal });
    return res;
  } finally {
    clearTimeout(timer);
  }
}

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

// ── Provider adapters ──────────────────────────────────────────────────────────

async function callGemini(model, apiKey, prompt, timeoutMs) {
  const url = `https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent?key=${apiKey}`;
  const res = await fetchWithTimeout(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ contents: [{ parts: [{ text: prompt }] }] }),
  }, timeoutMs);
  if (!res.ok) { const err = new Error(`HTTP ${res.status}`); err.status = res.status; throw err; }
  const data = await res.json();
  return data.candidates[0].content.parts[0].text;
}

async function callOpenAI(model, apiKey, prompt, timeoutMs) {
  const res = await fetchWithTimeout('https://api.openai.com/v1/chat/completions', {
    method: 'POST',
    headers: { 'Authorization': `Bearer ${apiKey}`, 'Content-Type': 'application/json' },
    body: JSON.stringify({ model, messages: [{ role: 'user', content: prompt }] }),
  }, timeoutMs);
  if (!res.ok) { const err = new Error(`HTTP ${res.status}`); err.status = res.status; throw err; }
  const data = await res.json();
  return data.choices[0].message.content;
}

async function callClaude(model, apiKey, prompt, timeoutMs) {
  const res = await fetchWithTimeout('https://api.anthropic.com/v1/messages', {
    method: 'POST',
    headers: {
      'x-api-key': apiKey,
      'anthropic-version': '2023-06-01',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ model, max_tokens: 1024, messages: [{ role: 'user', content: prompt }] }),
  }, timeoutMs);
  if (!res.ok) { const err = new Error(`HTTP ${res.status}`); err.status = res.status; throw err; }
  const data = await res.json();
  return data.content[0].text;
}

async function callMistral(model, apiKey, prompt, timeoutMs) {
  const res = await fetchWithTimeout('https://api.mistral.ai/v1/chat/completions', {
    method: 'POST',
    headers: { 'Authorization': `Bearer ${apiKey}`, 'Content-Type': 'application/json' },
    body: JSON.stringify({ model, messages: [{ role: 'user', content: prompt }] }),
  }, timeoutMs);
  if (!res.ok) { const err = new Error(`HTTP ${res.status}`); err.status = res.status; throw err; }
  const data = await res.json();
  return data.choices[0].message.content;
}

async function callGroq(model, apiKey, prompt, timeoutMs) {
  const res = await fetchWithTimeout('https://api.groq.com/openai/v1/chat/completions', {
    method: 'POST',
    headers: { 'Authorization': `Bearer ${apiKey}`, 'Content-Type': 'application/json' },
    body: JSON.stringify({ model, messages: [{ role: 'user', content: prompt }] }),
  }, timeoutMs);
  if (!res.ok) { const err = new Error(`HTTP ${res.status}`); err.status = res.status; throw err; }
  const data = await res.json();
  return data.choices[0].message.content;
}

async function callOllama(model, _apiKey, prompt, timeoutMs) {
  const res = await fetchWithTimeout('http://localhost:11434/api/generate', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ model, prompt, stream: false }),
  }, timeoutMs);
  if (!res.ok) { const err = new Error(`HTTP ${res.status}`); err.status = res.status; throw err; }
  const data = await res.json();
  return data.response;
}

const PROVIDERS = {
  gemini: callGemini,
  openai: callOpenAI,
  claude: callClaude,
  mistral: callMistral,
  groq: callGroq,
  ollama: callOllama,
};

// ── Main Client ────────────────────────────────────────────────────────────────

class FallbackClient {
  /**
   * @param {string} configPath - Path to config.json
   * @param {boolean} verbose   - Print fallback events to console
   */
  constructor(configPath = './config.json', verbose = true) {
    const raw = fs.readFileSync(path.resolve(configPath), 'utf-8');
    const cfg = JSON.parse(raw);
    this.models = cfg.models;
    this.retryLimit = cfg.retry_limit ?? 2;
    this.timeoutMs = (cfg.timeout_seconds ?? 30) * 1000;
    this.verbose = verbose;
  }

  _log(msg) {
    if (this.verbose) console.log(`[HermesFallback] ${msg}`);
  }

  /**
   * Attempt one model up to (retryLimit + 1) times.
   * @returns {Promise<string|null>} response text, or null if all attempts failed
   */
  async _tryModel(modelCfg) {
    const { provider, model } = modelCfg;
    const apiKey = resolveKey(modelCfg.api_key || '');
    const caller = PROVIDERS[provider];

    if (!caller) {
      this._log(`Unknown provider '${provider}', skipping.`);
      return null;
    }

    for (let attempt = 1; attempt <= this.retryLimit + 1; attempt++) {
      try {
        this._log(`Trying ${provider}/${model} (attempt ${attempt})...`);
        const result = await caller(model, apiKey, this._currentPrompt, this.timeoutMs);
        this._log(`✅ Success with ${provider}/${model}`);
        return result;

      } catch (err) {
        // Timeout (AbortError)
        if (err.name === 'AbortError') {
          this._log(`⏱ Timeout on ${provider}/${model} (attempt ${attempt})`);
          // fall through to retry

        } else if (err.status) {
          if (RETRYABLE_STATUS.has(err.status)) {
            this._log(`⚠️  HTTP ${err.status} on ${provider}/${model} (attempt ${attempt})`);
            if (attempt <= this.retryLimit) await sleep(2 ** (attempt - 1) * 1000);
          } else {
            this._log(`❌ Non-retryable HTTP ${err.status} on ${provider}/${model}`);
            return null; // skip this model immediately
          }

        } else {
          // Connection error or other
          this._log(`🔌 Error on ${provider}/${model}: ${err.message}`);
          // fall through to retry
        }
      }
    }

    this._log(`❌ All attempts exhausted for ${provider}/${model}, switching...`);
    return null;
  }

  /**
   * Send a prompt. Automatically falls back through all configured models.
   * @param {string} prompt
   * @returns {Promise<string>}
   * @throws {AllModelsFailedError}
   */
  async chat(prompt) {
    this._currentPrompt = prompt;

    for (const modelCfg of this.models) {
      const result = await this._tryModel(modelCfg);
      if (result !== null) return result;
    }

    throw new AllModelsFailedError(
      'All models failed (429 / 503 / Timeout). ' +
      'Check your API keys and quotas, or add more fallback models in config.json.'
    );
  }
}

module.exports = { FallbackClient, AllModelsFailedError };

// ── Quick test ─────────────────────────────────────────────────────────────────
if (require.main === module) {
  const client = new FallbackClient('./config.json');
  client.chat('Say hello in one sentence.')
    .then(res => console.log('Response:', res))
    .catch(err => console.error('Error:', err.message));
}
